lib/master.js

```diff
@@ -153,112 +156,108 @@ class Master extends BaseService {
-                            break;
-                        case 'ratelimiter_counters':
-                            return this._ratelimiter
-                                && this._ratelimiter.updateCounters(msg.value);
-                        default:
-                            this._logger.log('error/service-runner/master',
-                                `unknown message type received from worker ${msg.type}`);
-                    }
+            const workerExit = (worker) => {
+                if (this._shuttingDown || this._inRollingRestart) {
+                    return;
+                }
+
+                const info = {
+                    message: `worker died, restarting`,
+                    worker_pid: worker.process.pid,
+                    exit_code: worker.process.exitCode
+                };
+                if (this.workerStatusMap[worker.process.pid]
+                    && this.workerStatusMap[worker.process.pid].status) {
+                    info.status = this.workerStatusMap[worker.process.pid].status;
+                }
+                this._logger.log('error/service-runner/master', info);
+                delete this.workerStatusMap[worker.process.pid];
+                P.delay(Math.random() * 2000).then(() => {
+                    resolve(this._startWorker(worker.worker_id));
                 });
+            };
+
+            worker.on('exit', startupWorkerExit);
+            worker.on('message', (msg) => {
+                switch (msg.type) {
+                    case 'startup_finished':
+                        worker.removeListener('exit', startupWorkerExit);
+                        worker.on('exit', () => {
+                            workerExit(worker);
+                        });
+                        this._firstWorkerStarted = true;
+                        resolve(msg.serviceReturns);
+                        break;
+                    case 'heartbeat':
+                        this._saveBeat(worker);
+                        break;
+                    case 'service_status':
+                        this._onStatusReceived(worker, msg.status);
+                        break;
+                    case 'ratelimiter_counters':
+                        return this._ratelimiter
+                            && this._ratelimiter.updateCounters(msg.value);
+                    default:
+                        this._logger.log('error/service-runner/master',
+                            `unknown message type received from worker ${msg.type}`);
+                }
             });
-        } else {
-            return res;
-        }
+        });
+    }
+    // Fork a single worker, wait for it to start executing and set everything up,
+    // and then fork all the rest of the workers.
+    _startWorkers(workersToStart) {
+        return this._startWorker(1)
+        .then(firstReturn => P.map(
+            Array.from(new Array(workersToStart - 1), (val, i) => i + 2),
```

Comment:
I'm not entirely clear why (i) you are using the `(val, i)` form of the mapper function; and (ii) why you skip one index with `i + 2`

Reply:
(i) `new Array(workersToStart - 1)` with produce a `workersToStart - 1` element array of `undefined`, so to convert it to array of `2,3,4...` we need an index, the `val` will always be `undefined`

(ii) These indexes are used as the worker ID. eg, worker number 1, number 2 etc. WE're starting the first worker (number 1) sequentially, so here we want to generate an array of remaining IDs, eg [2,3,4,...,num_workers].